### ISSUE TITLE
Distributed build with /Zi can cause issues with debug info (MSVC)

### ISSUE THREAD

**Issue by john-blackburn** [fastbuild/fastbuild#493]
Consider a case where I have a main program `main.cpp` and a library consisting of several source files `func1.cpp`, `func2.cpp` etc in subdirectory `lib`. I build the library using MSVC like this

```
Compiler('msvc')
{
        .Root = '$VSBasePath$/VC/bin/amd64'
        .Executable = '$Root$/cl.exe'
        .ExtraFiles =
        {
           :
        }
        .CompilerFamily = 'msvc'
}

.Compiler           = 'msvc'
.CompilerOptions    = '"%1"'           // Input
                    + ' /Fo"%2"'       // Output
                    + ' /Zi'           // Debug in PDB file
                    + ' /c'            // Compile only
                    + ' /nologo'       // No compiler spam
                    + ' /W4'           // Warning level 4
                    + ' /WX'           // Warnings as errors
                    + ' /EHsc'         // Exceptions
                    + ' /FS'           // Force synchronous PDB writes
.Linker             = '$VSBasePath$\VC\bin\amd64\link.exe'
.LinkerOptions      = ' /OUT:"%2"'     // Output
                    + ' "%1"'          // Input
                    + ' /WX'           // Warnings as errors
                    + ' /NOLOGO'       // No linker spam
                    + ' /DEBUG'        // Keep debug info when linking
:
:
Library('mylib')
{
   .CompilerInputPath  = 'lib'
   .CompilerOutputPath = 'lib\out'
   .LibrarianOutput    = 'lib\out\funcs.lib'
   .CompilerOptions + ' /Fd"lib\out\funcs.pdb"'
}
```

Note that I have requested all the debug information will go into file `funcs.pdb` which contains debug info for the entire library.  So commands like the following will be issued
```
cl /c /Fo"lib\out\func1.obj" /Fd"lib\out\funcs.pdb" func1.cpp
cl /c /Fo"lib\out\func2.obj" /Fd"lib\out\funcs.pdb" func2.cpp
```

In order to make this work I need to add the `/FS` compiler flag so that writes to `funcs.pdb` are made synchronous.

Everything is fine on a single machine. But if I compile distributed, the PDB output is placed in files like this
```
lib\out\func1.obj.pdb
lib\out\func2.obj.pdb
```
Specifically, when the file was built remotely the debug information is put in a separate file such as `func1.obj.pdb`. When it was built locally it is compiled into the file that was requested, `funcs.pdb`

This behaviour is not as I requested! At link time, the linker may warn it can't find the debug information due to this being in the wrong files. This means the debug info will not be present on debugging. At the very least the creation of thousands of PDB files is likely to slow down linking.

This is a serious problem as it means it may be impossible to use the `Zi` option with FASTBuild when doing a distributed compile. `Z7` is 5 times slower to link that `Zi` in my testing. I'm getting link times of 15 mins in my project which makes development nearly impossible.

I attach a complete set of files to reproduce the problem. Just run `fbuild -dist -forceremote`.

[gather_pdb_error.zip](https://github.com/fastbuild/fastbuild/files/2414870/gather_pdb_error.zip)

**Comment by john-blackburn** [comment]
To actual see a link error replace line 80 in my attachment's `fbuild.bff` with 
```
   .CompilerOutputPath = 'lib\out\obj'
```
Then run `fbuild -dist -clean -forceremote`

you will get something like
```
C:\Users\john.blackburn\Documents\fastbuildtest\libs>fbuild -clean -dist -forceremote
Distributed Compilation : 4 Workers in pool
6> Obj: C:\Users\john.blackburn\Documents\fastbuildtest\libs\Out\main.obj
4> Obj: C:\Users\john.blackburn\Documents\fastbuildtest\libs\lib\out\obj\func8.obj
8> Obj: C:\Users\john.blackburn\Documents\fastbuildtest\libs\lib\out\obj\func9.obj
2> Obj: C:\Users\john.blackburn\Documents\fastbuildtest\libs\lib\out\obj\func7.obj
3> Obj: C:\Users\john.blackburn\Documents\fastbuildtest\libs\lib\out\obj\func5.obj
7> Obj: C:\Users\john.blackburn\Documents\fastbuildtest\libs\lib\out\obj\func2.obj
1> Obj: C:\Users\john.blackburn\Documents\fastbuildtest\libs\lib\out\obj\func3.obj
5> Obj: C:\Users\john.blackburn\Documents\fastbuildtest\libs\lib\out\obj\func6.obj
5> Obj: C:\Users\john.blackburn\Documents\fastbuildtest\libs\lib\out\obj\func1.obj
3> Obj: C:\Users\john.blackburn\Documents\fastbuildtest\libs\lib\out\obj\func4.obj
-> Obj: C:\Users\john.blackburn\Documents\fastbuildtest\libs\lib\out\obj\func6.obj <REMOTE: UK12847>
-> Obj: C:\Users\john.blackburn\Documents\fastbuildtest\libs\lib\out\obj\func5.obj <REMOTE: UK11920>
-> Obj: C:\Users\john.blackburn\Documents\fastbuildtest\libs\lib\out\obj\func3.obj <REMOTE: UK12847>
-> Obj: C:\Users\john.blackburn\Documents\fastbuildtest\libs\lib\out\obj\func7.obj <REMOTE: UK11411>
-> Obj: C:\Users\john.blackburn\Documents\fastbuildtest\libs\lib\out\obj\func2.obj <REMOTE: UK17192>
-> Obj: C:\Users\john.blackburn\Documents\fastbuildtest\libs\lib\out\obj\func9.obj <REMOTE: UK11920>
-> Obj: C:\Users\john.blackburn\Documents\fastbuildtest\libs\lib\out\obj\func8.obj <REMOTE: UK12847>
-> Obj: C:\Users\john.blackburn\Documents\fastbuildtest\libs\lib\out\obj\func1.obj <REMOTE: UK11411>
-> Obj: C:\Users\john.blackburn\Documents\fastbuildtest\libs\lib\out\obj\func4.obj <REMOTE: UK17192>
-> Obj: C:\Users\john.blackburn\Documents\fastbuildtest\libs\Out\main.obj <REMOTE: UK17192>
6> Lib: C:\Users\john.blackburn\Documents\fastbuildtest\libs\lib\out\funcs.lib
4> Exe: C:\Users\john.blackburn\Documents\fastbuildtest\libs\main.exe
4> funcs.lib(func1.obj) : warning LNK4099: PDB 'func1.obj.pdb' was not found with 'funcs.lib(func1.obj)' or at 'C:\Users\john.blackburn\Documents\fastbuildtest\libs\func1.obj.pdb'; linking object as if no debug info
LINK : error LNK1218: warning treated as error; no output file generated
4> Failed to build Exe (error 4099) 'C:\Users\john.blackburn\Documents\fastbuildtest\libs\main.exe'
FBuild: Error: BUILD FAILED: all
Time: 1.213s
```
Using dumpbin reveals the object file is looking for the PDB on the remote machine it was built on:
```
C:\Users\john.blackburn\Documents\fastbuildtest\libs\lib\out\obj>dumpbin /section:.debug$T /rawdata func1.obj
Microsoft (R) COFF/PE Dumper Version 14.00.24215.1
Copyright (C) Microsoft Corporation.  All rights reserved.


Dump of file func1.obj

File Type: COFF OBJECT

SECTION HEADER #3
.debug$T name
       0 physical address
       0 virtual address
      74 size of raw data
     5FF file pointer to raw data (000005FF to 00000672)
       0 file pointer to relocation table
       0 file pointer to line numbers
       0 number of relocations
       0 number of line numbers
42100040 flags
         Initialized Data
         Discardable
         1 byte align
         Read Only

RAW DATA #3
  00000000: 04 00 00 00 6E 00 15 15 24 F1 AD 48 50 9F E7 43  ....n...$ñ­HP.çC
  00000010: 8D 1C AC F9 9B 3A CB B6 01 00 00 00 43 3A 5C 55  ..¬ù.:Ë¶....C:\U
  00000020: 73 65 72 73 5C 42 48 41 4D 7E 31 2E 42 55 49 5C  sers\BHAM~1.BUI\
  00000030: 41 70 70 44 61 74 61 5C 4C 6F 63 61 6C 5C 54 65  AppData\Local\Te
  00000040: 6D 70 5C 2E 66 62 75 69 6C 64 2E 74 6D 70 5C 30  mp\.fbuild.tmp\0
  00000050: 78 30 30 30 30 30 30 30 30 5C 63 6F 72 65 5F 31  x00000000\core_1
  00000060: 30 30 34 5C 66 75 6E 63 31 2E 6F 62 6A 2E 70 64  004\func1.obj.pd
  00000070: 62 00 F2 F1                                      b.òñ

  Summary

          74 .debug$T
```

**Comment by ffulin** [comment]
I spent some time investigating this, and I understand what is going on.

**Summary of problem**

When FASTBuild builds on a remote host, and you want to use Zi debug info, FASTBuild has to make a pdb remotely. This pdb has to be returned to the local host and placed in a location where the link step will be able to find it to merge into the final pdb that is created during linking. It is this final pdb that is used for debugging, so as long as everything is resolved during linking, the final pdb contains complete debug info and is safe for debugging.

The issue you are seeing stems from how the linker looks for pdb files when linking a library into an executable. It looks in two locations:
 - next to the .lib being linked
 - in the original (full) path to the pdb as embedded in the obj when it was originally compiled

In the remote compilation case, naturally the full path is of no use as it's only valid on the remote machine. We therefore need the remotely built pdb files to be in the same directory as the lib when they are returned locally. When they are, as in your attachment without the modification, everything works fine. When they are not (after your edit), it fails to find them.

Sadly, it's not as simple as just making sure the lib and obj files end up in the same place, because obj files can end up in sub directories. You can reproduce the same problem by taking your original (working) zip, and moving any source file in a subdirectory; the pdb similarly ends up in a sub directory when returned from the remote machine and it fails during linking.

Originally, FASTBuild used to put all the obj files in the same directory (and therefore the pdb files), but this is problematic when your library contains multiple files with the same name. For this reason, for some time, FASTBuild has put the obj and accompanying pdb into subdirectories matching the source file hierarchy.

**A possible fix**
I think this problem can be fixed by making the obj and pdb files once again stored in the base directory when returned from the remote host, but with uniquified (containing hash of path) names to avoid conflicts when a project contains multiple files with the same name.  e.g. func1.0x12345678.obj and func1.0x87654321.obj

It would be useful for FASTBuild to also detect when you are using /Zi and do not have the .lib path set to be in the CompilerOutputPath folder, and output an error, since (as demonstrated by your repro case) that won't work correctly.  If you don't have /WX enabled, you may also not see the warning, which will result in non-obvious debugging problems!

**Why nobody noticed this before**
At first, I was a little surprised that nobody had run into or reported this problem before (including myself), but there are two reasons for it. Most users are either using:
 - /Z7 debug info (which avoids the entire pdb mess for the object files built remotely)
 - Unity files, which merge multiple cpp files together with the resulting obj and pdb being put in the lowest level directory (next to the lib)

In my case, I'm using both of those things in all cases, though historically i have used /Zi with Unity as well (see section about /Z7 slowness)

**Possible work-arounds**
I'm hopeful for a proper fix as described above, but in the short term, here are the possible work-arounds I can think of:
 - **use Unity** -  As noted above, that will avoid this problem. It has the added benefit of massively speeding up compilation as well.
 - **use Z7** - Although this is unusable in your case, it avoids the problem and has the added benefit of being faster (without link time hits) for many code bases (more details below)

I will leave this issue open to track this problem.

**Z7 slow link times**
I have seen this before, but only on Xbox One. At the time, it seems that whatever fork of the compiler was used in the SDK I was using it had some super weird behaviour not present in the regular Windows compiler at the same time. i.e. Z7 link times were great for Windows (x86 and x64) but awful for Xbox One. At the time, we switched Xbox One to Zi debug info. As we were using Unity, we didn't experience the problem we're talking about. Are you experiencing this on Xbox One by any chance? If so, its seems like such a massive regression should be treated like a bug, it might be good to open up a dialog with MS and see if this is something they can fix?

It's a real stab in the dark, but it might be worth experimenting with /Z7 and /FASTLINK if you haven't already. /FASTLINK leaves the debug info in the object files, which could potentially avoid whatever the slow path is in the linker. Naturally this would only be useful for local builds: you'd need your "Final" targets to use Zi, Z7 with slow link or perhaps disable distributed compilation so that you can store a pdb for offline use.

**Comment by john-blackburn** [comment]
Thanks very much for your detailed reply. The long link times I get are for both Windows and Xbox. I am using the same compiler/linker on both so I'd expect it to be slow on both if it's slow on one. I get the impression that Microsoft's link linker is much slower anyway than the one used by gcc or clang. Also MS's idea of using separate PDB files to store symbols makes things much more difficult for distributed builds and definitely increases the link time which seems to be dominated, not by building the EXE, but by building the PDB which is crazy really. gcc/clang does not have a separate symbol file and I think this is the better approach which might explain why it is faster. But we have to live with what Microsoft has done.

I have heard of `/debug:fastlink` and will give this a try. I get the impression that `/Z7` plus `/debug:fastlink` is the equivalent of what gcc does. I'd certainly prefer to use `/Z7` if I can get link times down. It's simpler and also allows me to use caching.

It seems to me that FASTBuild is silently changing my requested compile command to produce per-object PDBs. So, on the remote machine, instead of doing

```
cl /c /Zi /Fo"lib\out\func2.obj" /Fd"lib\out\funcs.pdb" func2.cpp
```
It's doing
```
cl /c /Zi /Fo"lib\out\func2.obj" /Fd"lib\out\obj\func2.obj.pdb" func2.cpp
```
Is that correct?

I think this is a mistake and FASTBuild should simply honour the command the user requested even on the remote machine. So it would generate `funcs.pdb` remotely and then, during the librarian stage, it would grab that file back across the network (giving it a different name maybe like `funcs.99.fb.pdb` where 99 is the number of the remote computer) and concatenate it to the file of the same name on the local computer.

When the linker runs on `funcs.lib`, the object file `func2.obj` inside will have a reference to `c:\path\on\remote\machine\funcs.pdb`. It can't find that so will look for `funcs.pdb` in the same directory as the `funcs.lib`. It's standard practice to put `funcs.lib` and `funcs.pdb`  in the same directory so this restriction would be harmless I think.

I think this is the best way to solve the problem. It also honours what the user initially requested. (I don't think FASTBuild should silently alter commands, it possible: the user should not get surprising behaviour when -dist is activated). It's likely to be faster at link time also as we have a few dozen per-library PDBs rather than a few thousands per-object PDBs: the remote computers do more of the work.

The above involves concatenating PDB files. If that's difficult, an alternative would be to silently change the command to
```
cl /c /Zi /Fo"lib\out\func2.obj" /Fd"lib\out\funcs.99.fb.pdb" func2.cpp
```
so now we have a partial, per-library PDB created by machine 99. It has a unique name and is in the same directory as the LIB file.

**Comment by ffulin** [comment]
**Manipulating args**

It is a key design principle of FASTBuild that does not change the command line, except where it's necessary for it to work.

Altering the command line in this case s required because we must uniquify the pdb names, and as the pdb name is referenced in the obj file, we cannot rename it after it's generated.

**Uniquification by object vs by host**

As we've discussed, we need the remotely built pdb to end up in the same directory as the lib and have a unique name; unique vs and locally generated pdb and unique against other hosts.  Additionally though, the pdb needs to be unique to the obj file. Consider:
 - host 99 builds obj A and returns it obj and funcs.pdb
 - disconnect/reconnect for a future build
 - host 99 builds obj B and returns it and the funcs.pdb

We now have conflicting pdbs and no way to merge them so we're back to square one. There is no mechanism I know of to merge pdb files (aside from the final link). The pdb file format is proprietary and not publicly documented. You can see that naming the file funcs.<hostname>.pdb doesn't help.

A logical next thought might be to try and at least make a given remote host return fewer pdb files overall, by sharing them somehow within the one build. Sadly, this is also unworkable. Consider:
 - host 99 builds obj A, B and C and returns funcs.pdb for all 3
 - disconnect/reconnect for a future build
 - host 99 builds obj A

Problems include:
 - We no longer have the funcs.pdb from before (can't keep forever filling up disks) so we've lost debug info for B & C
 - Even if we still had the pdb with debug info for B & C, we can't know it's still valid. Maybe b.obj changed but was built by another host. If we return the pdb with A, B & C we now have conflicting debug info
 - If we return after A, B and C are finished, how do we know there are not more objects to come? Objects are discovered dynamically during the build.
 - If we could somehoe know, what happens if we build A and B, but disconnect before building C and returning the pdb. We've now lost the debug info and have to go back and discard the objects A & B.

I think having more pdbs could have an impact on link time; it's certainly not going to help The question is whether it'll be a problem or not. For example, will it impact only the full link, or will it impact incremental scenarios, and in both cases by how much. Similarly, it could be interesting to note if /Debug:FASTLINK can help that case too (works with /Zi as well as /Z7). In the case I saw, it wasn't an issue: the combination of incremental linking and Unity meant it was still a good experience. (/Debug:FASTLINK didn't exist at that time).

I probably sound like a broken record, but I can't stress enough how much Unity can help in these situations. Unity frequently results in 10x speed ups even for just local builds. As it greatly reduces the amount of duplicate debug info that is generated (and must be de-duplicated at link-time), it often helps linking a great deal as well.

**/Z7 & /Debug:FASLINK**

If you do end up trying /Z7 with /Debug:FASTLINK, or even /Zi with /Debug:FASTLINK, either of those with or without Unity, I'd be interested to know your findings.

**Comment by john-blackburn** [comment]
The crucial point is you should gather the per-lib PDB back from a remote computer _only when it has finished all compiles from that library_. Then the PDB then has all the symbol information for all source files in that lib and you get them all in one transfer.

So the remote computer 99 issues commands like this
```
cl /c /Zi /Fo"lib\out\func1.obj" /Fd"lib\out\funcs.99.pdb" func1.cpp  --> send back func1.obj
cl /c /Zi /Fo"lib\out\func2.obj" /Fd"lib\out\funcs.99.pdb" func2.cpp  --> send back func2.obj
cl /c /Zi /Fo"lib\out\func2.obj" /Fd"lib\out\funcs.99.pdb" func3.cpp  --> send back func3.obj

Now funcs.99.pdb is complete (symbol info from `func1.cpp, func2.cpp, func3.cpp`). Now send it back to host.
```
Only when all source code for the `funcs.lib` library has been compiled does the host gather the entire `funcs.99.pdb` back. So the PDB is not transferred at the same time as the OBJ. I think this addresses most of the points you raised.

One way to achieve this in practice would be to lazily gather back PDB files just before they are needed. So you could gather all the remote PDBs back just before the link stage at which point they must all be complete. But I think its better to do it at the librarian stage for each library. At that point all source for that lib must have been build so the per-lib PDB on each remote is complete.

Then on the host machine you would end up with this directory:
```
funcs.lib
funcs.pdb (from local machine)
funcs.99.pdb
funcs.17.pdb
:
obj\func1.obj  --> points to funcs.pdb
obj\func2.obj  --> points to funcs.99.pdb
obj\func3.obj  --> points to funcs.17.pdb
:
```
All the PDB files and the LIB are present in the same directory so everything will work at link stage.

**Comment by ffulin** [comment]
The one pdb per remote host simply doesn't work in some situations. I tried to explain that in the previous reply, but perhaps not well enough (or maybe it was lost in the wall of text). :) In any case, let me try again.

Consider:
 - As part of a build, we need to compile a library "funcs" containing 3 object files (A, B and C)
 - Host 99 is given all 3 object files and builds them
 - Host 99 returns the 3 object files and the funcs.99.pdb as suggested
 - Everything so far is fine

Now consider:
 - You modify a.cpp, and need to rebuild A.obj
 - The object is built locally
 - We now have two conflicting versions of debug info for object A (one in locally built funcs.pdb and one in the original funcs.99.pdb)

Alternatively consider:
 - you modify a.cpp and need to rebuild A.obj
 - The object is built by host 99
 - Host 99 returns A.objc and funcs.99.pdb
 - funcs.99.pdb only contains debug info for A (the host can't keep data between builds)
 - we've lost debug info for B and C

There are many variations on these general scenarios that can play out, all resulting in missing or conflicting debug information which continues to get worse over multiple builds.

**Comment by john-blackburn** [comment]
Yes, thanks very much, I see that my suggestion won't work. I now think the right thing to do is group the source files together and compile each group as a unit with corresponding PDB for the whole group.

So for example consider you have files `func1.cpp` to `func10.cpp`. The user wishes to compile these into
```
funcs.lib
funcs.pdb
obj\func1.obj  \
obj\func2.obj  |
obj\func3.obj  +----group 1
obj\func4.obj  |
obj\func5.obj  /

obj\func6.obj  \
obj\func7.obj  |
obj\func8.obj  +-----group 2  
obj\func9.obj  |
obj\func10.obj /
```
We group the object files with N=5 files in each group. Let's say on first compile we have two workers, local computer and remote. The local computer compiles group 1 files like this
```
cl /c /Zi /Fo"obj\func1.obj" /Fd"funcs.1.pdb" func1.cpp
cl /c /Zi /Fo"obj\func2.obj" /Fd"funcs.1.pdb" func2.cpp
cl /c /Zi /Fo"obj\func3.obj" /Fd"funcs.1.pdb" func3.cpp
:
```
So instead of building `funcs.pdb` as the user requested we are building `funcs.1.pdb` (where 1 refers to the _group number_ [not the machine number]) but notice that this file contains multiple source symbol info and is in the same directory as `funcs.pdb` next to the lib file. So it's closer to what the user requested.

On the remote machine we run
```
cl /c /Zi /Fo"obj\func6.obj" /Fd"funcs.2.pdb" func6.cpp
cl /c /Zi /Fo"obj\func7.obj" /Fd"funcs.2.pdb" func7.cpp
cl /c /Zi /Fo"obj\func8.obj" /Fd"funcs.2.pdb" func8.cpp
:
```
The obj files are sent back as soon as they are build and the `funcs.2.pdb` file (2 referring to the group number) is sent back when it is complete (eg at librarian stage)

So now the user has, on his machine,
```
funcs.lib
funcs.1.pdb
funcs.2.pdb
obj\func1.obj  *
obj\func2.obj  
obj\func3.obj  *
obj\func4.obj  
obj\func5.obj  

obj\func6.obj  *
obj\func7.obj  *
obj\func8.obj    
obj\func9.obj  
obj\func10.obj *
```
Now let's say that the files shown (*) need to be recompiled. This time (to generalise) we compile group 1 files (1-5) on a remote computer and group 2 (6-10) on the local machine. Since symbol info for group 1 files is in `funcs.1.pdb` we must _send `funcs.1.pdb` to the remote computer_ (it already contains correct info for func2, 4 and 5). We then execute, on remote,
```
cl /c /Zi /Fo"obj\func1.obj" /Fd"funcs.1.pdb" func1.cpp
cl /c /Zi /Fo"obj\func3.obj" /Fd"funcs.1.pdb" func3.cpp
```
The obj files are send back as usual. When both the above are complete, `funcs.1.pdb` has been updated and can be sent back to the host.

On the local computer we have all the PDBs present so don't need to send anything. Just run
```
cl /c /Zi /Fo"obj\func6.obj" /Fd"funcs.2.pdb" func6.cpp
cl /c /Zi /Fo"obj\func7.obj" /Fd"funcs.2.pdb" func7.cpp
cl /c /Zi /Fo"obj\func10.obj" /Fd"funcs.2.pdb" func10.cpp
```
and `funcs.2.pdb` will be updated to the correct state.

So this method involves sending PDB files across the network to a remote machine then grabbing them back. This may be slow. Clearly if all the files in a group need to be recompiled there is no point sending the PDB to the remote. We could also have a rule that if more than, say, 50% need to be recompiled, don't send the PDB.

Your current method is effectively the above with N=1 except that you name the PDB files after the object files and intersperse them with object files. (which then causes trouble with linking)

This method is a bit like Unity in the sense that groups of files are effectively treated as a single file.

Do you think this would work?

**Comment by ffulin** [comment]
I don't think this approach is workable.

We don't have perfect knowledge of everything that needs to be built before we start building it in order to be able to split up tasks into groups like this. Tasks are discovered dynamically during the build so when we know we need to build one obj file we don't know what other obj files might also need building. Tasks can vary between builds as cpp files are added or removed as well, which would complicate keeping these groupings balanced.

Additionally, removing the atomicity of building individual obj files by grouping them into shared pdb files we send and retrieve from remote machines is impractical when we take into account that network connections can be lost during a build. Coordinating this would require us to rollback or undo compilation of objects in cases where we compiled 5 things but lost the connection before retrieving the shared pdb.

I still think the correct fix is to uniquify the remote pdb names and move them into the base folder to avoid the pathing issue.

In terms of helping users mitigate the performance impact of the Zi pdb files, I think the recommendation is still to use Unity files and switch to /Z7.

**Comment by john-blackburn** [comment]
I think you are right and probably adding distributed support for MSVC using /Zi and non-unity is not worth the effort. Creating 1 PDB per OBJ would probably be so slow as to obviate all benefits of distributed build anyway...

I would suggest simply not allowing distribution for /Zi. So on this page:

http://www.fastbuild.org/docs/features/distribution.html

Under "Compilation Settings" > "MSVC", add "/Z7 must be used" in addition to the other restrictions. Probably most people use caching and distribution so making both have the same restrictions is easier for users to remember. That would be slightly over restricted but at least it's simple to understand!

The fundamental problem is that Microsoft's per-lib PDBs in /Zi are fundamentally non-parallelizable so IMO just avoid the whole problem and insist on /Z7 for distributed, period! Then MSVC acts more like clang and gcc (the correct way!). The recent introduction of debug:fastlink makes MSVC even closer to clang and gcc, which is all to the good.

**Comment by ffulin** [comment]
While generally, /Z7 is overall much better, in some extreme cases, with certain code bases, it makes link times so long that it cannot be used. In those cases, it can still be useful to use /Zi with distribution and gain some overall improvement, even if it's not as optimal as one would hope. I've personally used this setup and it works, although as noted it relies on use of Unity to avoid the possible conflict.

Even when using Unity, there are options that allow files to be temporarily isolated from Unity to improve local iteration, and for those cases at least I think it's worth resolving the conflict mentioned above, so I'll continue to leave this issue open for now.

Even though I don't think we should remove /Zi support, I do think the documentation might be good to update to recommend Z7 over Zi though (and to suggest Unity for that matter).

**Comment by john-blackburn** [comment]
I wonder if there has been any developments on this issue? Perhaps a workaround would be to allow the user to specify a "flat" structure for created OBJ or LIB files. That way, the user can ensure the OBJ files sit next to the PDB file(s). Of course it would be up to the user to ensure no naming clash exists.

**Comment by mihaisebea** [comment]
Hi,
i think i'm having the same issues. 
Now ... i know about z7 but i's not really usuable for me as I have some big libraries and with z7 the lib file will be over 4GB and the linker will fail to link it (using a 64b linker yes).
Second, I know about unity files and I do (ab)use them but there are cases where certain tasks do require modifying certain cpp files and it;s too time consuming to remove them from the unity build.
Now, i noticed that msvc linker supports passing pdb files as well as object files.
I wonder if we can modify the LinkerNode to pass the full path of the pdbs at the same time with the objs (if Zi is used) and if this will solve the problem. 
I'll try give it a go on my side and see if it works.

**Comment by john-blackburn** [comment]
I recently wrote some notes to self about /Zi and distributed build. Hopefully this clarifies when the problems occur...

FASTBuild generally supports distributed build just by using the -dist flag (also set `FBUILD_BROKERAGE_PATH`, `Compiler()` and run fbuild worker on each machine)

However there are some restrictions relating distributed build using MSVC:

- /ZW cannot be used
- Precompiled headers cannot be distributed (but distributed build still seems to work, might be slow or bloated OBJ files?)
- /clr and /analyze cannot be distributed
- When linking static LIBs into an EXE or DLL and `.LinkerLinkObjects=false`, debug symbols *might* not be added if LIB was compiled with /Zi

In distributed build per-OBJ PDBs are generated remotely and put next to .OBJ files on master machine. (ie /Fd spec is overridden on remote machine).

If the LIB file is placed somewhere else (which can't always be avoided) when the linker runs it cannot find the PDB files that were built remotely so you don't get debug symbols (Linker will give a 4099 warning)

This is only a problem if linker input contains a Library (ObjectLists are OK)

**Workaround 1** Put lib source files in a single folder. Ensure LIB ends up in the same folder as the OBJ.

**Workaround 2** Use ` .LinkerLinkObjects=true`. Then the linker will ignore the LIB file and just link the OBJ files it is made of (But this may increase link time)

**Workaround 3** Use /Z7 in which case per-OBJ PDB files are not needed (debug symbols in OBJ)

**Workaround 4** Use Unity and make sure OBJ files are next to LIB files.

**Best practice:** Use /Z7 when doing distributed build with MSVC

If /Z7 causes slow link times, use the `/DEBUG:FASTLINK` linker option (VS2017 or later)

**Comment by ffulin** [comment]
If you are having problems with very large libs when using /Z7, some possible work around I can think of are below.  Full disclosure: I have not encountered this specific problem, so I don't know if this will solve it or shift the problem elsewhere, but they might be easy to try:

a) Don't make libs at all. Just directly link the .obj files. In some cases this can actually be faster as well. This is as simple as replacing your Library() with ObjectList(). It does change the rules for linking slightly, but in a way that if it's important, you almost certainly already know about.

b) Make several libs for each library.  Visual Studio requires that one project file = one library, but FASTBuild has no such restriction. You can arbitrarily put half of your files in MyLibrary1.lib and the other half in MyLibrary2 and then link both of those. If your library contains some top level folders you can split at that level.

You can potentially use the above two fixes for all your libraries, or just the problematic ones.

**Comment by claimred** [comment]
I've ran into same issues with /Zi and -dist.

- /Z7 with /DEBUG:FULL makes linking 7 times slower, /DEBUG:FASTLINK mitigates it but /DEBUG:FASTLINK also causes huge trouble with RelWithDebugInfo build, at least with vs2015. devenv.exe debugger regular crashes for instance. Looks like it got fixed in vs2017+.

- Got 4GB+ libraries that exceed coff size limit. Thanks for the advice, breaking them into multiple parts solved the issue. Although, parts themselves are quite big.

Guess I should try making precompiled headers lighter and use unity.

On a side note, tried linking with lld-link.exe, it's much slower than vs2015 link.exe on our objects file built by vs2015 cl.exe /Z7.

**Comment by john-blackburn** [comment]
@claimred I had exactly the same problem with FASTLINK and VS2015. It all seems to link fine but Visual Studio will crash when you try to step through the code. It seems FASTLINK was introduced in a VS2015 update but was very buggy and effectively useless. I think it's stable in VS2017 and 2019.

/Z7 can indeed be slower but it depends on the codebase. It's actually faster than Zi in my current codebase. Anyway Z7 + FASTLINK is the way to go for distributed build. It kind of makes MSVC act like gcc and clang have always acted. Other distributed build systems like Incredibuild and SN DBS also struggle with Zi, it's just a non-parallel way of doing things since multiple cl.exe are competing for one intermediate PDB file.

**Comment by claimred** [comment]
> Z7 is 5 times slower to link that Zi in my testing

@john-blackburn you've mentioned earlier in that issues that /z7 was 5 times slower in your testing. Was it different project, what changed? I'm trying to figure out changes needed to the codebase to make /z7 linking bearable.

**Comment by slonopotamus** [comment]
> I'm trying to figure out changes needed to the codebase to make /z7 linking bearable.

As was already said (https://github.com/fastbuild/fastbuild/issues/493#issuecomment-582484042), precompiled header is not used for remote compilation jobs. Instead, full header contents are preprocessed and included into compilation unit source, so *each* of your .obj files grows by the size of symbols that came through precompiled header.

So, either reduce size of precompiled header or reduce number of compilation units (i.e. enable unity build) or disable distributed build for  for those parts of your program that do not use unity build.

**Comment by john-blackburn** [comment]
@claimred yes, it's a different codebase. The new codebase is several DLLs so link time is less anyway and is similar using either Zi or Z7. The old project was 100 static libs which are linked into a final EXE. Link time was unbearable with normal Z7, but Z7 + FASTLINK was OK. So provided you have VS17 or VS19 FASTLINK is the solution.
